File: todoforai_edge/fuse/operations.py
"""FUSE filesystem operations for todoforai resources.

Port of fs/fuse/src/index.ts (the ops object inside createFuse).
Uses the fusepy/refuse pattern: methods return values or raise FuseOSError.
"""

import errno
import logging
import os
import stat
import threading
import time
from typing import Dict, Optional, Tuple

from .path_mapping import SCHEMES, is_root, is_scheme_root, path_to_uri
from .resource_client import ResourceClient

logger = logging.getLogger("todoforai-edge.fuse")


class ResourceOperations:
    """FUSE operations backed by the todoforai ResourceService."""

    def __init__(self, client: ResourceClient):
        self._client = client
        self._lock = threading.Lock()
        self._next_fd = 10
        # fd → (uri, buffer_or_None)
        self._open_files: Dict[int, Tuple[str, Optional[bytes]]] = {}

        self._uid = os.getuid()
        self._gid = os.getgid()
        self._now = time.time()

    # ------------------------------------------------------------------
    # Helpers
    # ------------------------------------------------------------------

    def _dir_stat(self) -> dict:
        return {
            "st_mode": stat.S_IFDIR | 0o755,
            "st_nlink": 2,
            "st_uid": self._uid,
            "st_gid": self._gid,
            "st_size": 0,
            "st_atime": self._now,
            "st_mtime": self._now,
            "st_ctime": self._now,
        }

    def _file_stat(self, size: int, mtime: float) -> dict:
        return {
            "st_mode": stat.S_IFREG | 0o644,
            "st_nlink": 1,
            "st_uid": self._uid,
            "st_gid": self._gid,
            "st_size": size,
            "st_atime": self._now,
            "st_mtime": mtime,
            "st_ctime": mtime,
        }

    def _raise(self, err: int):
        from ._compat import FuseOSError
        raise FuseOSError(err)

    # ------------------------------------------------------------------
    # FUSE callbacks
    # ------------------------------------------------------------------

    def getattr(self, path: str, fh=None):
        if is_root(path) or is_scheme_root(path):
            return self._dir_stat()

        uri = path_to_uri(path)
        if uri is None:
            self._raise(errno.ENOENT)

        try:
            meta = self._client.get_metadata(uri)
        except OSError as exc:
            self._raise(exc.errno or errno.ENOENT)

        if meta.is_directory:
            return self._dir_stat()

        mtime = (meta.created_at / 1000) if meta.created_at else self._now
        return self._file_stat(meta.size, mtime)

    def readdir(self, path: str, fh=None):
        entries = [".", ".."]

        if is_root(path):
            entries.extend(SCHEMES)
            return entries

        uri = path_to_uri(path)
        if uri is None:
            # scheme root without trailing content → use scheme://
            parts = [p for p in path.split("/") if p]
            if parts and parts[0] in SCHEMES:
                uri = f"{parts[0]}://"
            else:
                return entries

        try:
            items = self._client.list(uri)
            entries.extend(item.name for item in items)
        except OSError:
            pass  # return empty dir on error, same as TS version

        return entries

    def open(self, path: str, flags):
        uri = path_to_uri(path)
        if uri is None:
            self._raise(errno.ENOENT)

        with self._lock:
            fd = self._next_fd
            self._next_fd += 1
            self._open_files[fd] = (uri, None)

        return fd

    def read(self, path: str, size: int, offset: int, fh: int):
        with self._lock:
            entry = self._open_files.get(fh)
        if entry is None:
            return b""

        uri, buf = entry

        # Lazy fetch on first read (same pattern as TS version)
        if buf is None:
            try:
                content = self._client.fetch(uri)
                buf = content.data
            except OSError:
                return b""
            with self._lock:
                self._open_files[fh] = (uri, buf)

        return buf[offset:offset + size]

    def release(self, path: str, fh: int):
        with self._lock:
            self._open_files.pop(fh, None)
        return 0

    # ------------------------------------------------------------------
    # Phase-2 stubs (required by some FUSE consumers / tools)
    # ------------------------------------------------------------------

    def statfs(self, path: str):
        return {
            "f_bsize": 4096,
            "f_frsize": 4096,
            "f_blocks": 2**20,
            "f_bfree": 2**19,
            "f_bavail": 2**19,
            "f_files": 2**16,
            "f_ffree": 2**15,
            "f_favail": 2**15,
            "f_namemax": 255,
        }

    def access(self, path: str, amode: int):
        return 0

    def chmod(self, path: str, mode):
        return 0

    def chown(self, path: str, uid, gid):
        return 0

    def utimens(self, path: str, times=None):
        return 0

File: bun/src/fuse/operations.ts
/**
 * FUSE filesystem operations for todoforai resources.
 * Port of todoforai_edge/fuse/operations.py
 *
 * These callbacks are designed for fuse-native or any Node FUSE binding
 * that follows the { getattr, readdir, open, read, release } pattern.
 */

import { SCHEMES, isRoot, isSchemeRoot, pathToUri } from "./path-mapping.js";
import type { ResourceClient } from "./resource-client.js";

const S_IFDIR = 0o40000;
const S_IFREG = 0o100000;

export class ResourceOperations {
  private client: ResourceClient;
  private nextFd = 10;
  private openFiles = new Map<number, { uri: string; buf: Uint8Array | null }>();
  private uid: number;
  private gid: number;
  private now: number;

  constructor(client: ResourceClient) {
    this.client = client;
    this.uid = process.getuid?.() ?? 0;
    this.gid = process.getgid?.() ?? 0;
    this.now = Date.now() / 1000;
  }

  private dirStat() {
    return {
      mode: S_IFDIR | 0o755,
      nlink: 2,
      uid: this.uid,
      gid: this.gid,
      size: 0,
      atime: this.now,
      mtime: this.now,
      ctime: this.now,
    };
  }

  private fileStat(size: number, mtime: number) {
    return {
      mode: S_IFREG | 0o644,
      nlink: 1,
      uid: this.uid,
      gid: this.gid,
      size,
      atime: this.now,
      mtime,
      ctime: mtime,
    };
  }

  async getattr(path: string) {
    if (isRoot(path) || isSchemeRoot(path)) return this.dirStat();

    const uri = pathToUri(path);
    if (!uri) return null; // ENOENT

    const meta = await this.client.getMetadata(uri);
    if (meta.isDirectory) return this.dirStat();

    const mtime = meta.createdAt ? meta.createdAt / 1000 : this.now;
    return this.fileStat(meta.size, mtime);
  }

  async readdir(path: string): Promise<string[]> {
    const entries = [".", ".."];

    if (isRoot(path)) {
      entries.push(...SCHEMES);
      return entries;
    }

    let uri = pathToUri(path);
    if (!uri) {
      const parts = path.split("/").filter(Boolean);
      if (parts.length === 1 && SCHEMES.includes(parts[0] as any)) {
        uri = `${parts[0]}://`;
      } else {
        return entries;
      }
    }

    try {
      const items = await this.client.list(uri);
      entries.push(...items.map(i => i.name));
    } catch {}

    return entries;
  }

  open(path: string): number | null {
    const uri = pathToUri(path);
    if (!uri) return null; // ENOENT

    const fd = this.nextFd++;
    this.openFiles.set(fd, { uri, buf: null });
    return fd;
  }

  async read(path: string, size: number, offset: number, fd: number): Promise<Uint8Array> {
    const entry = this.openFiles.get(fd);
    if (!entry) return new Uint8Array(0);

    if (!entry.buf) {
      try {
        const content = await this.client.fetch(entry.uri);
        entry.buf = content.data;
      } catch {
        return new Uint8Array(0);
      }
    }

    return entry.buf.slice(offset, offset + size);
  }

  release(_path: string, fd: number) {
    this.openFiles.delete(fd);
  }

  statfs() {
    return {
      bsize: 4096,
      frsize: 4096,
      blocks: 2 ** 20,
      bfree: 2 ** 19,
      bavail: 2 ** 19,
      files: 2 ** 16,
      ffree: 2 ** 15,
      favail: 2 ** 15,
      namemax: 255,
    };
  }
}
